File: src/com/easy/ValidPalindrome_125.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class ValidPalindrome_125 {
    public static void main(String[] args) {
        ValidPalindrome_125 mTest = new ValidPalindrome_125();
        String s = "0P";// ".,";//"race a car";
        System.out.println(mTest.isPalindrome(s));
    }

    // 2019年7月25日
    // Use dual index comparison
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 3 ms, faster than 96.39%
    // Memory Usage: 37 MB, less than 99.95%
    public boolean isPalindrome(String s) {
        s = s.toUpperCase();
        int mIndexLeft = 0, mIndexRight = s.length() - 1;
        while (mIndexLeft <= mIndexRight) {
            char mHead = s.charAt(mIndexLeft);
            char mTail = s.charAt(mIndexRight);
            if (!checkAlphabet(mHead)) {
                mIndexLeft++;
            } else if (!checkAlphabet(mTail)) {
                mIndexRight--;
            } else {
                if (mHead != mTail) {
                    return false;
                }
                mIndexLeft++;
                mIndexRight--;
            }
        }
        return true;
    }

    // Can replace with build-in library Character.isLetterOrDigit(c)
    private boolean checkAlphabet(char c) {
        if (c >= 65 && c <= 90) { // Upper Case
            return true;
        } else if (c >= 97 && c <= 122) { // Lower Case
            return true;
        } else if (c >= 48 && c <= 57) { // Digit Case
            return true;
        } else {
            return false;
        }
    }

    // 2019年7月25日
    // Use reverse String comparison
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 16 ms, faster than 26.66%
    // Memory Usage: 38.6 MB, less than 79.91%
    public boolean isPalindrome2(String s) {
        s = s.toLowerCase().replaceAll("\\W+", "");
        return s.equals(new StringBuilder().append(s).reverse().toString());
    }
}

File: Golang_src/easy/validPalindrome_125.go
package easy

import "strings"

// Use two pointer:
// Time Complexity: O(len(s))
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 64.55%
// Memory Usage: 2.9 MB, less than 69.95%
func isPalindrome(s string) bool {
	size := len(s)
	left, right := 0, size-1
	s = strings.ToLower(s)

	for left <= right {
		for left < size && isAlphabet(s[left]) != true {
			left++
		}
		for right >= 0 && isAlphabet(s[right]) != true {
			right--
		}
		if left >= size || right < 0 {
			break
		}
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}
	return true
}

func isAlphabet(ch byte) bool {
	if ch >= 'a' && ch <= 'z' {
		return true
	}
	if ch >= '0' && ch <= '9' {
		return true
	}
	return false
}

// Use optimize two pointer:
// Time Complexity: O(len(s))
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 64.55%
// Memory Usage: 2.9 MB, less than 69.95%
func isPalindrome2(s string) bool {
	size := len(s)
	left, right := 0, size-1
	s = strings.ToLower(s)

	for left < right {
		for left < right && !isAlphabet2(s[left]) {
			left++
		}
		for left < right && !isAlphabet2(s[right]) {
			right--
		}
		if left > right {
			break
		}
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}
	return true
}

func isAlphabet2(ch byte) bool {
	return ch >= 'a' && ch <= 'z' || ch >= '0' && ch <= '9'
}
